File: server/Gate/Client/ClientManager.cs
//////////////////////////////////////////////////////////////////////////
// 
// 文件：server/Gate/Client/ClientManager.cs
// 作者：Xoen Xie
// 时间：2023/06/19
// 描述：客户端连接管理
// 说明：
//
//////////////////////////////////////////////////////////////////////////
#pragma warning disable CS8625, CS8618, CS8603

using XSF;

namespace GateClient
{
    public class ClientManager : IModule, INetHandler
    {
        private IConnection? m_Connection;

        private Client[] m_Clients;

        private int m_nTotal;

        private byte m_GlobalKey = 0;

        private INetPacker m_ClientPakcer;

        public ClientManager()
        {
            m_Clients = new Client[XSFCore.Config.GateMaxCount];
            m_ClientPakcer = new LocalPacker();
        }

        public override void DoRegist()
        {
            XSFCore.SetMessageExecutor((ushort)XsfPbid.CMSGID.CltGtHandshake, new Executor_Clt_Gt_Handshake());
            XSFCore.SetMessageExecutor((ushort)XsfPbid.CMSGID.CltGtHeartbeat, new Executor_Clt_Gt_Heartbeat());
        }

        public override bool Start()
        {
            m_Connection = XSFCore.Listen(m_ClientPakcer, this, (int)XSFCore.Server.Ports[(int)EP.Client]);
            if(m_Connection == null)
                return false;

            return true;
        }

        public override void OnStartClose()
        {
            for(int i = 0; i < m_Clients.Length; i ++)
            {
                if(m_Clients[i] != null && m_Clients[i].ID > 0)
                    m_Clients[i].Disconnect((int)XsfPb.DisconnectReason.ServerDown);
            }
        }

        public override ModuleRunCode OnCloseCheck()
        {
            if(m_nTotal > 0)
                return ModuleRunCode.Wait;

            return ModuleRunCode.OK;
        }

        internal void Delete(Client c)
        {
            int nIndex = c.m_CID.ID;
            if(nIndex >= m_Clients.Length)
            {
                Serilog.Log.Error($"clientManager.Delete error, nIndex:{nIndex} >= m_Clients.Length:{m_Clients.Length}, client id={c.ID} [{c.CID.Gate}-{c.CID.ID}-{c.CID.Key}]");
                return;
            }

            if(c == m_Clients[nIndex])
            {
                m_Clients[nIndex].Release();
                m_nTotal --;
            }
            else
            {
                Serilog.Log.Error($"clientManager.Delete error, c {c.ID} == m_Clients[nIndex] {m_Clients[nIndex].ID}");
            }
        }

        public void OnAccept(IConnection connection) 
        {
            Serilog.Log.Information("clientManager OnAccept ...");
            if(m_nTotal >= XSFCore.Config.GateMaxCount)
            {
                connection.Close();
                return;
            }

            for(int i = 0; i < m_Clients.Length; i ++)
            {
                if(m_Clients[i] == null)
                {
                    Client client = new Client();
                    client.Create(this, connection);
                    client.m_CID.Gate = XSFCore.Server.SID.Index;
                    client.m_CID.ID = (ushort)i;
                    client.m_CID.Key = m_GlobalKey ++;
                    client.m_nID = ClientID.GetID(client.m_CID);

                    Serilog.Log.Information("clientManager OnAccept new ... id=" + client.ID);

                    connection.DoStart(client);

                    m_Clients[i] = client;

                    m_nTotal ++;

                    return;
                }
                else
                {
                    if(m_Clients[i].ID <= 0)
                    {
                        Client client = m_Clients[i];
                        client.Create(this, connection);
                        client.m_CID.Gate = XSFCore.Server.SID.Index;
                        client.m_CID.ID = (ushort)i;
                        client.m_CID.Key = m_GlobalKey ++;
                        client.m_nID = ClientID.GetID(client.m_CID);
                        connection.DoStart(client);
                        m_nTotal ++;

                        Serilog.Log.Information("clientManager OnAccept get old ... id=" + client.ID);

                        return;
                    }
                }
            }

            connection.Close();
        }

        public Client GetClient(uint nClientID)
        {
            var cid = ClientID.GetCID(nClientID);
            if(cid.ID >= m_Clients.Length)
            {
                Serilog.Log.Error($"ClientManager.GetClient cid.ID:{cid.ID} >= m_Clients.Length:{m_Clients.Length}");
                return null;
            }

            if(m_Clients[cid.ID] != null && m_Clients[cid.ID].ID == nClientID && m_Clients[cid.ID].IsHandshake)
            {
                return m_Clients[cid.ID];
            }

            return null;
        }

        public void Broadcast(byte[] data)
        {
            for(int i = 0; i < m_Clients.Length; i ++)
            {
                if(m_Clients[i] != null && m_Clients[i].IsHandshake)
                {
                    m_Clients[i].SendData(data);
                }
            }
        }

        public void OnConnected(IConnection connection) 
        {
            Serilog.Log.Error("ClientManager.OnConnected error call");
        }
        
        public void OnError(IConnection connection, NetError nErrorCode) 
        {
            Serilog.Log.Error("ClientManager.OnError error call");
        }
        
        public void OnRecv(IConnection connection, IMessage message, ushort nMessageID, uint nRawID, byte[]? data) 
        {
            Serilog.Log.Error("ClientManager.OnRecv error call");
        }

        public static ClientManager Instance
        {
            get; private set;
        }

        public static void CreateModule()
        {
            Instance = new ClientManager();

            ModuleInit init = new ModuleInit();
            init.ID = (int)ModuleID.Client;
            init.Name = "ClientManager";

            XSFCore.Server.AddModule(Instance, init);
        }
    }
}

File: server-cpp/source/gate/server/src/client/ClientManager.cpp
//////////////////////////////////////////////////////////////////////////
//
// 文件：source/gate/server/src/client/ClientManager.cpp
// 作者：Xoen Xie
// 时间：2020/08/05
// 描述：client 管理
// 说明：
//
//////////////////////////////////////////////////////////////////////////
#include "ClientManager.h"
#include "DMessageInc.h"
#include "Client.h"
#include "ClientExecutors.h"
#include "LocalPacker.h"
#include "Server.h"


Client * ClientManager::New(void)
{
    if( m_nTotal >= m_nSize )
    {
        XSF_WARN("ClientManager::New m_nCount[%u] >= m_nSize[%u]", m_nTotal, m_nSize);
        return nullptr;
    }

    Client * pNewClient = nullptr;
    uint32 nID = 0xFFFFFFFF;

    for( uint32 i = 0; i < m_nSize; ++ i )
    {
        if( m_Clients[i] == nullptr )       // 当前位置为空，直接可用
        {
            Client * pClient = new Client();
            m_Clients[i] = pClient;
            pNewClient = pClient;
            nID = i;
            break;
        }
        else
        {
            Client * pClient = m_Clients[i];
            if( pClient->m_SID.ID == 0 )         // 如果ID为0，则表示当前Client已经被释放掉
            {
                pNewClient = pClient;
                nID = i;
                break;
            }
        }
    }

    if( pNewClient != nullptr )
    {
        pNewClient->m_SID.C.gate = XSFCore::GetServer()->GetSID()->S.index;
        pNewClient->m_SID.C.id = nID;
        pNewClient->m_SID.C.key = m_nGlobalKey ++;
        m_nTotal ++;

        return pNewClient;
    }

    XSF_ERROR("ClientManager::New can't found an empty client, m_nCount[%u], m_nSize[%u]", m_nTotal, m_nSize);

    return nullptr;
}

void ClientManager::Delete( Client * pClient )
{
    if( pClient->m_SID.ID == 0 )
    {
        XSF_ERROR("ClientModule::Delete pClient->m_ID.ID == 0");
        return;
    }

    uint32 nIndex = pClient->m_SID.C.id;
    if( m_Clients[nIndex] != pClient )
    {
        XSF_WARN("ClientModule::Delete m_Clients[nIndex] != pClient");
        return;
    }

    pClient->m_SID.ID = 0;
    m_nTotal --;
}

Client * ClientManager::GetClient( uint32 nClientID )
{
    SID sid;
    sid.ID = nClientID;

    if( sid.C.gate != XSFCore::GetServer()->GetSID()->S.index )
    {
        XSF_ERROR("ClientModule::Get sid.C.gate[%u] != XSFCore::GetServer()->GetID()->S.index[%u]", sid.C.gate, XSFCore::GetServer()->GetSID()->S.index);
        return nullptr;
    }
    
    uint32 nIndex = sid.C.id;
    if( m_Clients[nIndex] == nullptr )
        return nullptr;

    if( m_Clients[nIndex]->m_SID.C.key != sid.C.key )
        return nullptr;

    if( !m_Clients[nIndex]->m_bHandshake )
        return nullptr;

    return m_Clients[nIndex];
}

void ClientManager::Broadcast( const byte * pData, uint32 nDataLen )
{
    for( uint32 i = 0; i < m_nSize; ++ i )
    {
        Client * pClient = m_Clients[i];
        if( pClient != nullptr && pClient->m_bHandshake )
        {
            pClient->SendData(pData, nDataLen);
        }
    }
}

bool ClientManager::Init( ModuleInit * pInit )
{
    IServer * pServer = XSFCore::GetServer();
    auto pConfig = pServer->GetConfig();

    m_nSize = pConfig->GateMaxCount;
    if( m_nSize <= 0 )
        m_nSize = 1024;

    m_Clients = new Client * [m_nSize];
    memset( m_Clients, 0, sizeof(Client *)*m_nSize);

    m_pPakcer = new LocalPacker();

    return IModule::Init(pInit);
}

void ClientManager::Release(void)
{
    CONNECTION_CLOSE(m_pConnection);

    if( m_Clients != nullptr )
    {
        for( uint32 i = 0; i < m_nSize; ++ i )
        {
            if( m_Clients[i] != nullptr )
            {
                delete m_Clients[i];
            }
        }

        delete [] m_Clients;
        m_Clients = nullptr;
    }

    XSF_DELETE(m_pPakcer);
    
    delete this;
}

void ClientManager::DoRegist()
{
    XSFCore::SetMessageExecutor(xsf_pbid::Clt_Gt_Handshake, new Executor_Clt_Gt_Handshake());
    XSFCore::SetMessageExecutor(xsf_pbid::Clt_Gt_Heartbeat, new Executor_Clt_Gt_Heartbeat());
}

bool ClientManager::Start()
{
    m_pConnection = XSFCore::CreateConnection(this, m_pPakcer);
    if(m_pConnection == nullptr)
        return false;

    auto ports = XSFCore::GetServer()->GetPorts();
    return m_pConnection->Listen(ports[EP_Client]);
}

void ClientManager::OnStartClose(void)
{
    for( uint32 i = 0; i < m_nSize; ++ i )
    {
        if( m_Clients[i] != nullptr && m_Clients[i]->m_SID.ID > 0 )
        {
            m_Clients[i]->Disconnect(xsf_pb::DisconnectReason::ServerDown);
        }
    }
}

uint8 ClientManager::OnCloseCheck()
{
    if(m_nTotal > 0)
        return ModuleRunCode_Wait;

    return ModuleRunCode_OK;
}


// 接受到连接
INetHandler * ClientManager::OnAccept(IConnection * pIncomingConn)
{
    FUNCTION_ENTER(0);
    Client * pNewClient = New();
    if(pNewClient == nullptr)
    {
        return nullptr;
    }

    pNewClient->Create(pIncomingConn);
    const SID * pID = pNewClient->GetSID();
    XSF_WARN("client login, ip=[%s:%u], client=[%u %u-%u]", pIncomingConn->GetRemoteIP(), pIncomingConn->GetRemotePort(), pID->ID, pID->C.id, pID->C.key);

    return pNewClient;

    FUNCTION_EXIT(0);
}

// 已连接到远程主机
void ClientManager::OnConnected( IConnection * pConection )
{
    XSF_ERROR("ClientManager::OnConnected illegal call");
}

// 连接出错或断开
void ClientManager::OnError( IConnection * pConection, uint32 nErrorCode )
{
    XSF_ERROR("ClientManager::OnError error:%u", nErrorCode);
}

// 收到数据
void ClientManager::OnRecv(IConnection * pConection, DataResult * pResult)
{
    XSF_ERROR("ClientManager::OnRecv illegal call");
}

void SetClientManager(void)
{
    ModuleInit * pInit = new ModuleInit();
    pInit->nModuleID = GateModule_Client;
    strcpy(pInit->sName, "ClientManager");

    XSFCore::GetServer()->AddModule(ClientManager::Instance(), pInit);
}